### PR TITLE
Time Series Preprocessing - Imputing & Exogenous model extension

### LINKED ISSUE #1752: Add imputation capability in time series module
Add ability to impute missing values in a dataset

Link to [sktime imputer](https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.transformations.series.impute.Imputer.html)

### LINKED ISSUE #2275: Enable exogenous variable support for CdsDt models in Time Series
**Is your feature request related to a problem? Please describe.**
Reduced Regression Models are nothing but regression models under the hood. These should be able to support exogenous variables. 

**Describe the solution you'd like**
This was not enabled until now since the tags might have indicated otherwise. But based on this analysis, it seems that they do support exogenous variables. Hence, enable them in PyCaret now.

https://github.com/alan-turing-institute/sktime/issues/2206

### PR DESCRIPTION
**Pull request by ngupta23**
## Related Isuse or bug

Closes #1752
Closes #2275 

#### Describe the changes you've made

- Implements the framework for preprocessing in time series and uses imputation as a pipe cleaning example
- Implements exogenous support for CdsDt reduced regression models per #2275

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Unit tests have been added

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: pycaret/time_series/forecasting/oop.py [hunk 1]
```diff
@@ -1406,16 +1492,79 @@ def update_fit_kwargs_with_fh_from_cv(fit_kwargs: Optional[Dict], cv) -> Dict:
...
+        """Extracts and returns the final model from the pipeline. The pipeline
+        may be of type ForecastingPipeline or TransformedTargetForecaster or of
+        type sktime BaseForecaster. These conditions are handled appropriately"""
+
+        if self.transform_exogenous and self.transform_target:
+            #### Exogenous Variables and targets are being transformed ----
+            # The last step is a `TransformedTargetForecaster`
+            # pipeline.steps[-1] is the TransformedTargetForecaster
+            # pipeline.steps[-1][1].steps[-1] is the final model
+            final_forecaster_only = pipeline.steps[-1][1].steps[-1][1]
```
Comment (ngupta23) [inline pycaret/time_series/forecasting/oop.py:1539]: Change to `steps_` to get the fitted model (last step)

**Comment by ngupta23** [comment]
@tvdboom @Yard1 Can you please review these changes. I have made them according to what we discussed. 
Also, I am not sure why 1 test is passing on Linux but failing on Windows (related to Reg/Classification preprocessing). @tvdboom Could you have a look at this please? Thanks.

image

**Comment by ngupta23** [comment]
> Also, I am not sure why 1 test is passing on Linux but failing on Windows (related to Reg/Classification preprocessing). @tvdboom Could you have a look at this please? Thanks.

@tvdboom, I think it is some numerical instability issue on Windows. I ran it multiple times on my Windows machine and it passed sometimes and failed sometimes. I just reduced the threshold from 1 to 0.9999. I hope that is acceptable.

**Review by tvdboom — APPROVED** [review]

File: pycaret/internal/preprocess/time_series/forecasting/preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            raise ValueError(
+                f"{target} Imputation Type '{numeric_imputation}' is not of allowed type."
+            )
+
+        if target:
+            self.pipe_steps_target.extend([("numerical_imputer", num_estimator)])
+        else:
+            self.pipe_steps_exogenous.extend([("numerical_imputer", num_estimator)])
+
+    # def _transformation(self, transformation_method):
```
[the remark this answers is not in the document]
Comment (ngupta23) [reply pycaret/internal/preprocess/time_series/forecasting/preprocessor.py:54]: Thanks for approving. I have only started with imputation right now until the approval. I will add the rest shortly after this is merged.

**Pull request merged**